src/parameters.jl

```diff
@@ -172,6 +174,82 @@ function reconstruct(obj, values::NamedTuple{keys}) where {keys}
     return setproperties(obj, patch)
 end
 
+"""
+    $SIGNATURES
+
+Convenience macro for creating "parameterized" `struct`s. Fields can be marked as parameters with `@param`, e.g:
+
+```julia
+@parameterized Base.@kwdef struct Foo{T}
+    @param x::T = 1.0
+    y::T = NaN
+end
+```
+where here `x` will be treated as a parameter field and `y` will not. Parameters will be then automatically
+generated in a corresponding `parameters(::Foo)` method definition. Additional parameter attributes can be
+supplied as keywords after the parameter, e.g. `@param p::T = 0.5 bounds=UnitInterval()` or
+`@param p::T = 0.5 (bounds=UnitInterval(), constant=false)`.
+"""
+macro parameterized(expr)
+    # process struct definition
+    ## first declare local variables for capturing metadata
+    paraminfo = []
+    lastdoc = nothing
+    typename = nothing
+    ## prewalk (top-down) over expression tree; note that there is some danger of infinite
+    ## recursion with prewalk since it will descend into the returned expression. However,
+    ## this is not a problem here since we only deconstruct/reduce the @param expressions.
+    new_expr = MacroTools.prewalk(expr) do ex
+        # case 1: struct definition (top level)
+        if isa(ex, Expr) && ex.head == :struct
+            # type signature is in second argument;
+            # use namify to extract type name (discard type parameters)
+            typename = namify(ex.args[2])
+            ex
+        # case 2: parameterized field definition
+        elseif @capture(ex, @param fieldname_::FT_ = defval_; attrs__) || @capture(ex, @param fieldname_::FT_; attrs__)
+            # use last seen docstring as description, if present
+            desc = isnothing(lastdoc) ? "" : lastdoc
+            attrs = if isempty(attrs)
+                (;)
+            elseif length(attrs) == 1
+                # handle both singleton attr=value syntax as well as named tuple syntax (attr1=value1, attr2=value2, ...)
+                attrs[1].head == :tuple ? eval(attrs[1]) : eval(:(($(attrs...),)))
+            else
+                error("invalid syntax for parameter attributes: $(QuoteNode(attrs))")
+            end
+            # then reset to nothing to prevent duplication
+            lastdoc = nothing
+            push!(paraminfo, (name=fieldname, type=FT, desc=desc, attrs...))
+            :($fieldname::$FT = $defval)
+        # case 3: non-parameterized field definition
+        elseif @capture(ex, fieldname_::FT_ = defval_) || @capture(ex, fieldname_::FT_)
+            # reset lastdoc variable to prevent confusing docstrings between parameteter and non-parameter fields
+            lastdoc = nothing
+            ex
+        # case 4: field documentation
+        elseif @capture(ex, docs_String)
+            # store in lastdoc field
+            lastdoc = docs
```

Comment:
This technically would allow

```julia
    "doc string"
    # empty line here
    a::T
```

no? Because I think DocStringExtensions.jl doesn't allow that?!

Reply:
Yes, it would allow that, but I don't think you should do it?

Reply:
Yeah no. Just thought out loud here. I think it's fine don't see this causing problems